File: src/player/features/androidAuto/contentStyle.ts
/**
 * PLAYER-267: the Android Auto content-style contract.
 *
 * The native side (ContentStyle.kt) maps a JS MediaItem.extras carrying these keys onto the
 * androidx.media MediaConstants content-style bundle, which media3's MediaLibrarySession surfaces to
 * Android Auto. These mirror MediaConstants by VALUE (verified against androidx.media:media:1.7.0):
 *   DESCRIPTION_EXTRAS_VALUE_CONTENT_STYLE_LIST_ITEM = 1, _GRID_ITEM = 2,
 *   _CATEGORY_LIST_ITEM = 3, _CATEGORY_GRID_ITEM = 4.
 *
 * GUAU's CONTENT_STYLE enum (baseTemplate.utils.ts) emits LIST = '1', GRID = '2' as the extras values —
 * which equal CONTENT_STYLE_HINT.LIST / .GRID below, so the native string->int parse is the identity on
 * the numeric value.
 */

/** The extras keys the JS layer puts on a MediaItem to request a content-style. */
export const CONTENT_STYLE_EXTRA_KEYS = {
  /** This item's own presentation style. */
  contentStyle: 'contentStyle',
  /** Presentation style for this item's browsable children. */
  childrenBrowsableContentStyle: 'childrenBrowsableContentStyle',
  /** Presentation style for this item's playable children. */
  childrenPlayableContentStyle: 'childrenPlayableContentStyle',
  /** Grouping title for this item's children. */
  groupTitle: 'groupTitle',
} as const;

/** Content-style hint values (mirror MediaConstants ints, verified media3 1.1.1 / androidx.media 1.7.0). */
export enum CONTENT_STYLE_HINT {
  LIST = 1,
  GRID = 2,
  CATEGORY_LIST = 3,
  CATEGORY_GRID = 4,
}

/** Parse a GUAU content-style extras value ("1"/"2") to a hint int, or undefined if unrecognised. */
export const parseContentStyleHint = (value: string | number | undefined): CONTENT_STYLE_HINT | undefined => {
  const n = typeof value === 'string' ? Number.parseInt(value, 10) : value;
  return n === CONTENT_STYLE_HINT.LIST || n === CONTENT_STYLE_HINT.GRID ||
         n === CONTENT_STYLE_HINT.CATEGORY_LIST || n === CONTENT_STYLE_HINT.CATEGORY_GRID
    ? (n as CONTENT_STYLE_HINT)
    : undefined;
};

File: android/src/main/java/com/brentvatne/exoplayer/androidauto/ContentStyle.kt
package com.brentvatne.exoplayer.androidauto

import android.os.Bundle
import androidx.media.utils.MediaConstants

/**
 * PLAYER-267: bridge GUAU's content-style intent (carried in JS MediaItem.extras by the PLAYER-273
 * templatesToMediaLibrary mapper) into the androidx.media content-style bundle keys that media3's
 * MediaLibrarySession surfaces to Android Auto (via MediaUtils.convertToMediaDescriptionCompat, which
 * copies MediaMetadata.extras -> MediaDescriptionCompat.extras — verified media3 1.1.1).
 *
 * GUAU CONTENT_STYLE values: LIST = "1", GRID = "2"  (baseTemplate.utils.ts:96-99) — these map directly
 * onto MediaConstants.DESCRIPTION_EXTRAS_VALUE_CONTENT_STYLE_LIST_ITEM (1) / _GRID_ITEM (2).
 */
object ContentStyle {

    // GUAU extras keys (must match mediaLibrary.utils.ts:18-21).
    private const val KEY_CONTENT_STYLE = "contentStyle"
    private const val KEY_CHILDREN_BROWSABLE = "childrenBrowsableContentStyle"
    private const val KEY_CHILDREN_PLAYABLE = "childrenPlayableContentStyle"
    private const val KEY_GROUP_TITLE = "groupTitle"

    /** Map a GUAU style string ("1"/"2") to the media3 int hint, or null if unrecognised. */
    private fun toHint(value: String?): Int? = when (value) {
        "1" -> MediaConstants.DESCRIPTION_EXTRAS_VALUE_CONTENT_STYLE_LIST_ITEM // 1
        "2" -> MediaConstants.DESCRIPTION_EXTRAS_VALUE_CONTENT_STYLE_GRID_ITEM // 2
        else -> null
    }

    /**
     * Build the media3 content-style Bundle from GUAU's extras map. Returns an empty Bundle when no
     * content-style keys are present (safe to setExtras unconditionally).
     */
    fun toBundle(extras: Map<String, String?>): Bundle {
        val b = Bundle()
        toHint(extras[KEY_CONTENT_STYLE])?.let {
            b.putInt(MediaConstants.DESCRIPTION_EXTRAS_KEY_CONTENT_STYLE_SINGLE_ITEM, it)
        }
        toHint(extras[KEY_CHILDREN_BROWSABLE])?.let {
            b.putInt(MediaConstants.DESCRIPTION_EXTRAS_KEY_CONTENT_STYLE_BROWSABLE, it)
        }
        toHint(extras[KEY_CHILDREN_PLAYABLE])?.let {
            b.putInt(MediaConstants.DESCRIPTION_EXTRAS_KEY_CONTENT_STYLE_PLAYABLE, it)
        }
        extras[KEY_GROUP_TITLE]?.takeIf { it.isNotEmpty() }?.let {
            b.putString(MediaConstants.DESCRIPTION_EXTRAS_KEY_CONTENT_STYLE_GROUP_TITLE, it)
        }
        return b
    }

    /** True if the bundle carries no content-style key (used to decide whether to persist it). */
    fun isEmpty(b: Bundle): Boolean = b.isEmpty
}
